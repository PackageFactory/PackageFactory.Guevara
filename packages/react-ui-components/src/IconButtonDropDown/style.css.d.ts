export interface IStyleCss {
  'wrapper': string;
  'wrapper__btn': string;
  'wrapper__btnModeIcon': string;
  'wrapper__btnIcon': string;
  'wrapper__dropDown': string;
  'wrapper__dropDown--isOpen': string;
  'wrapper__dropDownItem': string;
}

export const locals: IStyleCss;
