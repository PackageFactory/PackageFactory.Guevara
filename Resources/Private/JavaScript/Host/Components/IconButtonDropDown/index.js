import React, {Component, PropTypes} from 'react';
import mergeClassNames from 'classnames';
import {executeCallback} from 'Host/Abstracts/';
import {service} from 'Shared/';
import Icon from 'Host/Components/Icon/';
import Button from 'Host/Components/Button/';
import style from './style.css';

const {logger} = service;

export default class IconButtonDropDown extends Component {
    static propTypes = {
        // The icon key which is always displayed.
        icon: PropTypes.string.isRequired,

        // You can pass an modeIcon which displays the current selected item in a leaner way.
        // Modify this prop via listening to the `onItemSelect` propType.
        modeIcon: PropTypes.string.isRequired,

        // Style related propTypes.
        className: PropTypes.string,
        isDisabled: PropTypes.bool,

        // Child items of the DropDown.
        children: PropTypes.node.isRequired,

        // Interaction related propTypes.
        onClick: PropTypes.func.isRequired,
        onItemSelect: PropTypes.func.isRequired
    }

    constructor(props) {
        super(props);

        this.mouseHoldTimeout = null;
        this.state = {isOpened: false};
    }

    render() {
        const {
            className,
            isDisabled,
            icon,
            modeIcon
        } = this.props;
        const classNames = mergeClassNames({
            [style.wrapper]: true,
            [className]: className && className.length
        });
        const dropDownClassNames = mergeClassNames({
            [style.wrapper__dropDown]: true,
            [style['wrapper__dropDown--isOpen']]: this.state.isOpened
        });

        return (
            <div className={classNames} onMouseLeave={this.onMouseLeave.bind(this)}>
                <Button
                    isDisabled={isDisabled}
                    className={style.wrapper__btn}
                    onMouseDown={this.createHoldTimeout.bind(this)}
                    onClick={this.onClick.bind(this)}
                    >
                    <Icon icon={modeIcon} className={style.wrapper__btn__modeIcon} />
                    <Icon icon={icon} />
                </Button>
                <div className={dropDownClassNames}>
                    {this.renderChildren()}
                </div>
            </div>
        );
    }

    renderChildren() {
        return this.props.children
            .map(child => typeof child === 'function' ? child() : child)
            .filter(child => child)
            .map((child, index) => (
                <a
                    className={style.wrapper__dropDown__item}
                    onClick={this.onItemSelect.bind(this, child.props.dropDownId)}
                    key={index}
                    >
                    {child}
                </a>
            ));
    }

    createHoldTimeout() {
        this.mouseHoldTimeout = setTimeout(() => {
            this.openDropDown();
        }, 200);
    }

    cancelHoldTimeout() {
        clearTimeout(this.mouseHoldTimeout);
    }

    onClick() {
        const {isOpened} = this.state;

        this.cancelHoldTimeout();

        if (!isOpened) {
            executeCallback(this.props.onClick);
        }
    }

    onMouseLeave() {
        this.cancelHoldTimeout();
        this.closeDropDown();
    }

    onItemSelect(ref) {
        const {onItemSelect} = this.props;

        if (!ref) {
            logger.error('Please specify a ref="" prop on your "IconButtonDropDown" children.');
        }

        if (onItemSelect) {
            onItemSelect(ref);
        }

        this.closeDropDown();
    }

    openDropDown() {
        this.setState({isOpened: true});
    }

    closeDropDown() {
        this.setState({isOpened: false});
    }
}
IconButtonDropDown.defaultProps = {
    modeIcon: 'long-arrow-right',
    isDisabled: false
};
