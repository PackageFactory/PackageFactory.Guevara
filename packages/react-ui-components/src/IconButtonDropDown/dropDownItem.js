import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';

class DropDownItem extends PureComponent {
    static propTypes = {
        /**
         * The handler to call when clicking on the item of the DropDown.
         */
        onClick: PropTypes.func.isRequired,

        /**
         * The ID to reference the clicked item in the `onClick` handker.
         */
        id: PropTypes.string.isRequired,

        /**
         * The children to render within the anchor.
         */
        children: PropTypes.element.isRequired
    }

    handleClick = () => {
        const {onClick, id} = this.props;
        onClick(id);
    };

    render() {
        const {children, ...rest} = this.props;

        return (
            <a
                {...rest}
                role="button"
                onClick={this.handleClick}
                >
                {children}
            </a>
        );
    }
}

export default DropDownItem;
