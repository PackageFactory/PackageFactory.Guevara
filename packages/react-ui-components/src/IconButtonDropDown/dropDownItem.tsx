import React, {PureComponent} from 'react';

type DropDownItemId = string;

interface IDropDownItemProps {
    // TODO: think about declaring currying the function with the id in the parent component/container
    /**
     * The handler to call when clicking on the item of the DropDown.
     */
    readonly onClick: (id: DropDownItemId) => void;

    /**
     * The ID to reference the clicked item in the `onClick` handker.
     */
    readonly id: DropDownItemId;

    /**
     * The children to render within the anchor.
     */
    readonly children: React.ReactNode;

    readonly className: string;
}

class DropDownItem extends PureComponent<IDropDownItemProps> {
    public render(): JSX.Element {
        const {children, ...rest} = this.props;

        return (
            <a
                {...rest}
                role="button"
                onClick={this.handleClick}
                >
                {children}
            </a>
        );
    }

    private readonly handleClick = () => {
        this.props.onClick(this.props.id);
    }
}

export default DropDownItem;
