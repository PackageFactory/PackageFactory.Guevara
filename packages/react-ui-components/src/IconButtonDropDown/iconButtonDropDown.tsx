import mergeClassNames from 'classnames';
import React, {PureComponent} from 'react';

import {IButtonProps} from '../Button/button';
import DropDownItem from './dropDownItem';

type DropDownId = string;

interface IIconButtonDropDownTheme {
    readonly 'wrapper': string;
    readonly 'wrapper__btn': string;
    readonly 'wrapper__btnModeIcon': string;
    readonly 'wrapper__dropDown': string;
    readonly 'wrapper__dropDown--isOpen': string;
    readonly 'wrapper__dropDownItem': string;
    readonly [key: string]: string;
}

interface IIconButtonDropDownProps {
    /**
     * The key of the Icon which is always displayed.
     */
    readonly icon: string;

    // TODO: be more specific
    // TODO: use render prop
    /**
     * Children to be rendered inside the DropDown.
     */
    readonly children: ReadonlyArray<React.ReactElement<{
        readonly dropDownId: string;
    }>>;

    /**
     * You can pass an modeIcon which displays the current selected item in a leaner way.
     * Modify this prop via listening to the `onItemSelect` propType.
     */
    readonly modeIcon: string;

    /**
     * Will be called once the user clicks on the wrapper if
     * it is NOT opened right meow.
     */
    readonly onClick: () => void;

    /**
     * Will be called once the user selects an item.
     * You should specify an `dropDownId` prop on each child,
     * which will then be propagated to this handler as the first
     * and only argument.
     */
    readonly onItemSelect: (id: DropDownId) => void;

    /**
     * An optional css theme to be injected.
     */
    readonly theme?: IIconButtonDropDownTheme;

    // TODO: This feels strange. We should just import the component classes here.
    // also those interfaces should be exposed (export interface ...)
    /**
     * Static component dependencies which are injected from the outside (index.js)
     */
    readonly IconComponent: React.ComponentClass<{
        readonly icon: string;
        readonly className?: string;
    }>;

    readonly ButtonComponent: React.ComponentClass<IButtonProps>;

    /**
     * An optional `className` to attach to the wrapper.
     */
    readonly className?: string;

    /**
     * Controls the whole components disabled state.
     */
    readonly isDisabled: boolean;

    // we can remove this if we should import and use the _actual_ component
    /**
     * Props which are propagated to the <Button> component.
     */
    readonly directButtonProps: Partial<IButtonProps>;
}

interface IIconButtonDropDownState {
    readonly isOpen: boolean;
    readonly mouseHoldTimeout: number | undefined;
    readonly mouseHoverTimeout: number | undefined;
}

export default class IconButtonDropDown extends PureComponent<IIconButtonDropDownProps, IIconButtonDropDownState> {
    constructor(props: IIconButtonDropDownProps) {
        super(props);

        this.state = {
            isOpen: false,
            mouseHoldTimeout: undefined,
            mouseHoverTimeout: undefined,
        };
    }

    public static readonly defaultProps: Partial<IIconButtonDropDownProps> = {
        directButtonProps: {},
        isDisabled: false,
        modeIcon: 'long-arrow-right',
    };

    public render(): JSX.Element {
        const {
            IconComponent,
            ButtonComponent,
            className,
            isDisabled,
            icon,
            modeIcon,
            theme,
            directButtonProps,
            children
        } = this.props;
        const {isOpen} = this.state;
        const finalClassName = mergeClassNames(className, theme!.wrapper);
        const dropDownClassNames = mergeClassNames(
            theme!.wrapper__dropDown,
            {
                [theme!['wrapper__dropDown--isOpen']]: isOpen,
            }
        );

        const {size, ...rest} = directButtonProps;

        return (
            <div className={finalClassName} onMouseLeave={this.handleMouseLeave}>
                <ButtonComponent
                    {...rest}
                    style="clean"
                    hoverStyle="clean"
                    aria-haspopup="true"
                    className={theme!.wrapper__btn}
                    isDisabled={isDisabled}
                    onMouseDown={this.handleHoldTimeout}
                    onMouseEnter={this.handleHoverTimeout}
                    onFocus={this.handleHoverTimeout}
                    onClick={this.handleClick}
                    size={size!}
                >
                    <IconComponent icon={modeIcon} className={theme!.wrapper__btnModeIcon}/>
                    <IconComponent icon={icon} className={theme!.wrapper__btnIcon}/>
                </ButtonComponent>
                <div className={dropDownClassNames} aria-hidden={isOpen ? 'false' : 'true'}>
                    {children.map((child) => (
                        <DropDownItem
                            key={child.props.dropDownId}
                            className={theme!.wrapper__dropDownItem}
                            onClick={this.handleItemSelected}
                            id={child.props.dropDownId}
                            >
                            {child}
                        </DropDownItem>
                    ))}
                </div>
            </div>
        );
    }

    private readonly createHoldTimeout = () => {
        this.setState({
            mouseHoldTimeout: window.setTimeout(() => this.openDropDown(), 200),
        });
    }

    private readonly cancelHoldTimeout = () => {
        const {mouseHoldTimeout} = this.state;

        if (mouseHoldTimeout) {
            clearTimeout(mouseHoldTimeout);
            this.setState({
                mouseHoldTimeout: undefined,
            });
        }
    }

    private readonly createHoverTimeout = () => {
        this.setState({
            mouseHoverTimeout: window.setTimeout(() => this.openDropDown(), 700),
        });
    }

    private readonly cancelHoverTimeout = () => {
        const {mouseHoverTimeout} = this.state;

        if (mouseHoverTimeout) {
            clearTimeout(mouseHoverTimeout);
            this.setState({
                mouseHoverTimeout: undefined,
            });
        }
    }

    private readonly handleHoverTimeout = () => this.createHoverTimeout();

    private readonly handleHoldTimeout = () => this.createHoldTimeout();

    private readonly handleClick = () => {
        const {isOpen} = this.state;

        this.cancelHoldTimeout();
        this.cancelHoverTimeout();

        if (!isOpen) {
            this.props.onClick();
        }
    }

    private readonly handleMouseLeave = () => {
        this.cancelHoldTimeout();
        this.cancelHoverTimeout();
        this.closeDropDown();
    }

    private readonly handleItemSelected = (ref: DropDownId) => {
        this.props.onItemSelect(ref);
        this.closeDropDown();
    }

    private readonly openDropDown = () => {
        this.setState({isOpen: true});
    }

    private readonly closeDropDown = () => {
        this.setState({isOpen: false});
    }
}
