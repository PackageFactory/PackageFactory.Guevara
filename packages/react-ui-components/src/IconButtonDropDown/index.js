import {themr} from '@friendsofreactjs/react-css-themr';
import identifiers from '../identifiers';
import style from './style.css';
import IconButtonDropDown from './iconButtonDropDown';

const ThemedIconButtonDropDown = themr(identifiers.iconButtonDropDown, style)(IconButtonDropDown);

// TODO: why here?
//
// Dependency injection
//
import injectProps from './../_lib/injectProps';
import Icon from './../Icon';
import Button from './../Button';

export default injectProps({
    IconComponent: Icon,
    ButtonComponent: Button
})(ThemedIconButtonDropDown);
