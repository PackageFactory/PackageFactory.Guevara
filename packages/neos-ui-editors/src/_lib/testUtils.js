import {PureComponent, Children} from 'react';
import PropTypes from 'prop-types';
import {SynchronousRegistry, SynchronousMetaRegistry} from '@neos-project/neos-ui-extensibility/src/registry';

const globalRegistry = new SynchronousMetaRegistry();

// i18n Registry
class FakeI18NRegistry extends SynchronousRegistry {
    translate(key) {
        return key;
    }
}
globalRegistry.set('i18n', new FakeI18NRegistry());

// Data Source DataLoader
export const MockDataSourceDataLoader = {
    _currentPromise: null,
    _currentPromiseResolveFn: null,

    // API in tests
    reset() {
        this._currentPromise = null;
    },

    resolveCurrentPromise(result) {
        this._currentPromiseResolveFn(result);
        return this._currentPromise;
    },

    // API of DataLoader
    resolveValue() {
        if (!this._currentPromise) {
            this._currentPromise = new Promise(resolve => {
                this._currentPromiseResolveFn = resolve;
            });
        }
        return this._currentPromise;
    }
};
globalRegistry.set('dataLoaders', new SynchronousRegistry());
globalRegistry.get('dataLoaders').set('DataSources', MockDataSourceDataLoader);

// wrapper
export class WrapWithMockGlobalRegistry extends PureComponent {
    static propTypes = {
        children: PropTypes.element.isRequired
    };

    static childContextTypes = {
        globalRegistry: PropTypes.object.isRequired,
        configuration: PropTypes.object.isRequired
    };

    getChildContext() {
        const configuration = {};
        return {globalRegistry, configuration};
    }

    render() {
        return Children.only(this.props.children);
    }
}
