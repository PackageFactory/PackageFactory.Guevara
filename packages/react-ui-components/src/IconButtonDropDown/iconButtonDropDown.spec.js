import sinon from 'sinon';
import {createShallowRenderer, createStubComponent} from './../_lib/testUtils.js';
import IconButtonDropDown from './iconButtonDropDown.js';
import DropDownItem from './dropDownItem.js';

const Icon = createStubComponent();
const Button = createStubComponent();
const defaultProps = {
    icon: 'barGeneralIcon',
    children: [
        createStubComponent()({dropDownId: 'foo1'}),
        createStubComponent()({dropDownId: 'foo2'})
    ],
    modeIcon: 'fooModeIcon',
    onClick: () => null,
    onItemSelect: () => null,
    theme: {
        'wrapper': 'baseWrapperClassName',
        'wrapper__btn': 'baseBtnClassName',
        'wrapper__btnModeIcon': 'baseBtnIconClassName',
        'wrapper__dropDown': 'baseDropDownClassName',
        'wrapper__dropDown--isOpen': 'baseDropDownOpenClassName',
        'wrapper__dropDownItem': 'baseDropDownItemClassName'
    },
    IconComponent: Icon,
    ButtonComponent: Button
};
const shallow = createShallowRenderer(IconButtonDropDown, defaultProps);

test('should initialize with a state of "{isOpen: false}".', () => {
    const dd = shallow();

    expect(dd.state('isOpen')).toBe(false);
});
test('should render a "div" node with the themes "wrapper" className.', () => {
    const dd = shallow();

    expect(dd.type()).toBe('div');
    expect(dd.hasClass('baseWrapperClassName')).toBeTruthy();
});
test('should render the "className" prop if passed.', () => {
    const dd = shallow({
        className: 'barClassName'
    });

    expect(dd.hasClass('barClassName')).toBeTruthy();
});
test('should render a "ButtonComponent" component with style="clean" and aria-haspopup="true" prop and a className which matches the themes "wrapper__btn".', () => {
    const dd = shallow();
    const btn = dd.find(Button);

    expect(btn.prop('style')).toBe('clean');
    expect(btn.prop('aria-haspopup')).toBe('true');
    expect(btn.hasClass('baseBtnClassName')).toBeTruthy();
});
test('should render a "ButtonComponent" component which reflects the "isDisabled" prop.', () => {
    const dd = shallow({
        isDisabled: 'fooDisabled'
    });
    const btn = dd.find(Button);

    expect(btn.prop('isDisabled')).toBe('fooDisabled');
});
test('should set the "isOpen" state to "true" when pressing the "ButtonComponent" for more than 200 ms.', () => {
    const dd = shallow();
    const btn = dd.find(Button);

    return new Promise(resolve => {
        btn.simulate('mouseDown');

        setTimeout(() => {
            expect(dd.state('isOpen')).toBe(true);
            resolve();
        }, 300);
    });
});
test('should abort the setting of the "isOpen" state to "true" when pressing and afterwards clicking on the "ButtonComponent" within the 200 ms.', () => {
    const dd = shallow();
    const btn = dd.find(Button);

    return new Promise(resolve => {
        btn.simulate('mouseDown');

        btn.simulate('click');

        setTimeout(() => {
            expect(dd.state('isOpen')).toBe(false);
            resolve();
        }, 300);
    });
});
test('should call the "onClick" prop when clicking on the "ButtonComponent".', () => {
    const props = {
        onClick: sinon.spy()
    };
    const dd = shallow(props);
    const btn = dd.find(Button);

    btn.simulate('click');

    expect(props.onClick.calledOnce).toBeTruthy();
});

test('should render two "IconComponent"s within the "ButtonComponent".', () => {
    const dd = shallow();
    const btn = dd.find(Button);
    const icons = btn.find(Icon);

    expect(icons.length).toBe(2);
});
test('should propagate the "modeIcon" prop to the first "IconComponent".', () => {
    const dd = shallow();
    const btn = dd.find(Button);
    const icon = btn.find(Icon).at(0);

    expect(icon.prop('icon')).toBe('fooModeIcon');
});
test('should propagate the "icon" prop to the second "IconComponent".', () => {
    const dd = shallow();
    const btn = dd.find(Button);
    const icon = btn.find(Icon).at(1);

    expect(icon.prop('icon')).toBe('barGeneralIcon');
});

test('should render a "DropDownItem" for each passed child and propagate the "dropDownId" to it as "id".', () => {
    const dd = shallow();
    const items = dd.find(DropDownItem);

    expect(items.length).toBe(2);
    expect(items.at(0).prop('id')).toBe('foo1');
    expect(items.at(1).prop('id')).toBe('foo2');
});
test('should call the "onItemSelect" prop when clicking on a "DropDownItem".', () => {
    const props = {
        onItemSelect: sinon.spy()
    };
    const dd = shallow(props);
    const item = dd.find(DropDownItem).at(0);

    item.simulate('click');

    expect(props.onItemSelect.calledOnce).toBeTruthy();
});
