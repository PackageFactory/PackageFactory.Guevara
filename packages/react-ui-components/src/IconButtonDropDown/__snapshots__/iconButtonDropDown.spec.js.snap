// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IconButtonDropDown/> should render correctly. 1`] = `
<div
  className="baseWrapperClassName"
  onMouseLeave={[Function]}
>
  <Component
    aria-haspopup="true"
    className="baseBtnClassName"
    hoverStyle="clean"
    isDisabled={false}
    onClick={[Function]}
    onFocus={[Function]}
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    style="clean"
  >
    <Component
      className="baseBtnIconClassName"
      icon="fooModeIcon"
    />
    <Component
      icon="barGeneralIcon"
    />
  </Component>
  <div
    aria-hidden="true"
    className="baseDropDownClassName"
  >
    <DropDownItem
      className="baseIconButtonDropDownClassName"
      id="foo1"
      key="0/.0"
      onClick={[Function]}
    >
      <div
        dropDownId="foo1"
        style={Object {}}
      />
    </DropDownItem>
    <DropDownItem
      className="baseIconButtonDropDownClassName"
      id="foo2"
      key="1/.1"
      onClick={[Function]}
    >
      <div
        dropDownId="foo2"
        style={Object {}}
      />
    </DropDownItem>
  </div>
</div>
`;
